File: Vision_Artificial/practica6/practica6.m
    %practica n6 
%Programa que clasifica un vector respecto a n clases 
clc
clear
close all
format rational
% Solicitar al usuario que ingrese el número de clases
while true
    num_cl = input('Ingresa el numero de clases a generar: ');
    if (num_cl <= 0 || num_cl > 49)
        disp('NO es posible generar más de 49 clases, por favor ingresa un número válido.')
    else
        break
    end
end


%matriz para guardar las clases 
matrices = cell(1,num_cl);

while true
    num_de_objetos = input('Ingresa el numero de objetos por clase: ');
    if num_de_objetos <= 0
        disp('El numero de objetos debe ser mayor que 0.')
    else
        break
    end
end


for i = 1:num_cl    
    
    while true
        fprintf('Ingresa la dispersion en x para la clase %d\n ',i);
        disper_x = input(': ');
        fprintf('Ingresa la dispersion en y para la clase %d\n ',i);
        disper_y = input(': ');
        fprintf('Ingresa la dispersion en z para la clase %d\n', i);
        disper_z = input(': ');
        if (disper_x <= 0 || disper_y <= 0 || disper_z <= 0)
            disp('La dispersion debe de ser diferente de 0.');
        else   
            break
        end
    end
   
    fprintf('Ingresa el centroide de x para la clase %d\n ',i);
    centro_x = input(":");
    fprintf('Ingresa el centroide de y para la clase %d\n',i);
    centro_y = input(":");
    fprintf('Ingresa el centroide de z para la clase %d\n',i);
    centro_z = input(":");

    matrices{i} = [(centro_x + randn(1, num_de_objetos)*disper_x); ...
                   (centro_y + randn(1, num_de_objetos)*disper_y); ...
                   (centro_z + randn(1, num_de_objetos)*disper_z)];
end 
   
op = 0 ;
while (op ~= -1)
    %pedimos el vector al usuario
    vx=input('dame el valor de la coord en x=');
    vy=input('dame el valor de la coord en y=');
    vz=input('dame el valor de la coord en z=');
    vector=[vx;vy;vz];

        medias = cell(1,num_cl);
    distancias = zeros(1, num_cl);
    while true
        disp("Que modelo deseas usar para calcular las distancias a las clases");
        disp("1.Distancia Euclidiana.")
        disp("2.Mohalanobis.")
        modelo = input("Tu opcion: ");
        if modelo ~= 1 && modelo ~= 2
            disp('Opción no válida, por favor elige 1 o 2.');
        else    
            break
        end
    end
    vector1 = cell(1,num_cl);
    %obtenemos la media de cada clase
    for j = 1:num_cl
        medias{j} = mean(matrices{j}, 2);
    end

    if modelo == 1
            % Calcular la distancia de cada vector a las medias de las clases
            disp("Has escogido la distancia euclidiana. ")
            for j = 1:num_cl
                distancias(j) = norm(vector - medias{j});
            end
    else 
            %genera el vector con la suma entre el numero de objetos
            %totales
            
            for i = 1:num_cl
                disp("Has escogido la distancia Mohalanobis. ")
                matriz_clase = matrices{i};
                media_clase = medias{i};
                matriz_cov = cov(matriz_clase');
                sigma_inv = inv(matriz_cov);
                diff_vector = vector - media_clase;
                distancia_cad = (diff_vector' * sigma_inv) * diff_vector;
                distancias(i) = sqrt(distancia_cad);

            end 
 
    end

    % Determinar la clase más cercana
    [~, clase_asignada] = min(distancias);
    minimo = min(distancias);
    if minimo > 400
        disp(['El vector no pertenece a ninguna clase debido a que ' num2str(minimo)]);
        disp('es mayor al umbral de 100 puntos para poder pertenecer a alguna clase')
    else
        disp(['El vector pertenece a la clase: ' num2str(clase_asignada)]);
        disp(['Con una distancia de ' num2str(minimo)])
    end
    
    %para poder graficar
    leyendas = cell(1,num_cl);
    figure;
    hold on;

    %graficamos las clsaes 
    for i = 1: num_cl
        x_coords = matrices{i}(1,:);
        y_coords = matrices{i}(2,:);
        z_coords = matrices{i}(3,:);
        %crea el color de manear aleatoria
        color_rgb = rand(1,3);

        plot3(x_coords,y_coords,z_coords,'o','Color',color_rgb, "MarkerSize",8,"lineWidth",1.5);
        leyendas{i} = ['Clase ' num2str(i)];
    end
    leyendas{num_cl + 1} = ['Vector'];



    plot3(vector(1,:),vector(2,:),vector(3,:),'go','MarkerSize',10,'MarkerFaceColor','g')
    xlabel('Eje X');
    ylabel('Eje Y');
    zlabel('Eje Z');
    title('Clasificación de vectores por clase');
    legend(leyendas);
    grid on;
    view(3);
    hold off;

    
    op = input('Deseas continuar con el programa (-1 para salir): ');
end 
disp('Fin de programa, ahi nos vemos. ');

File: Vision_Artificial/practica6/practica6.py
import numpy as np
import matplotlib.pyplot as plt

# PROGRAMA QUE CLASIFICA UN VECTOR RESPECTO A N CLASES (Euclidiana o Mahalanobis)

# --- 1. SOLICITAR PARÁMETROS DE CLASIFICACIÓN ---

def obtener_parametro(mensaje, validacion):
    """Función auxiliar para manejar la entrada de usuario y la validación."""
    while True:
        try:
            valor = input(mensaje)
            # Evaluar la entrada si es necesario (para números, principalmente)
            if 'num_clases' in validacion:
                 valor = int(valor)
                 if 1 <= valor <= 49:
                     return valor
                 else:
                     print('NO es posible generar más de 49 clases, por favor ingresa un número válido (1-49).')
            elif 'num_objetos' in validacion:
                valor = int(valor)
                if valor > 0:
                    return valor
                else:
                    print('El numero de objetos debe ser mayor que 0.')
            elif 'dispersión' in validacion:
                valor = float(valor)
                if valor > 0:
                    return valor
                else:
                    print('La dispersión debe ser mayor que 0.')
            elif 'centroide' in validacion or 'coordenada' in validacion:
                 return float(valor)
            elif 'modelo' in validacion:
                 valor = int(valor)
                 if valor in [1, 2]:
                     return valor
                 else:
                     print('Opción no válida, por favor elige 1 (Euclidiana) o 2 (Mahalanobis).')
            else:
                return valor

        except ValueError:
            print("Entrada no válida. Por favor, ingresa un número.")

# Solicitar el número de clases
num_cl = obtener_parametro('Ingresa el numero de clases a generar: ', 'num_clases')

# Solicitar el número de objetos por clase
num_de_objetos = obtener_parametro('Ingresa el numero de objetos por clase: ', 'num_objetos')

# Lista para guardar las clases
matrices = []
dispersiones = [] # Se usa para guardar las dispersiones (disper_x, disper_y) si son necesarias

# --- 2. GENERACIÓN DE CLASES ---

for i in range(num_cl):
    print(f"\n--- CLASE {i + 1} ---")
    
    # Solicitar la dispersión (desviación estándar) para la clase actual
    while True:
        disper_x = obtener_parametro(f'Ingresa la dispersion en x para la clase {i+1}: ', 'dispersión')
        disper_y = obtener_parametro(f'Ingresa la dispersion en y para la clase {i+1}: ', 'dispersión')
        if disper_x > 0 and disper_y > 0:
            break
        
    dispersiones.append((disper_x, disper_y))

    # Solicitar el centroide de la clase actual
    centro_x = obtener_parametro(f'Ingresa el centroide de x para la clase {i+1}: ', 'centroide')
    centro_y = obtener_parametro(f'Ingresa el centroide de y para la clase {i+1}: ', 'centroide')

    # Generación de los puntos de la clase (distribución normal)
    x_coords = centro_x + np.random.randn(1, num_de_objetos) * disper_x
    y_coords = centro_y + np.random.randn(1, num_de_objetos) * disper_y
    
    # Apilamos las coordenadas para formar una matriz 2xN
    clase = np.vstack([x_coords, y_coords])
    matrices.append(clase)

# --- 3. CICLO PRINCIPAL DE CLASIFICACIÓN ---

op = 0
while op != -1:
    print("-" * 40)
    
    # Pedimos el vector al usuario
    vx = obtener_parametro('dame el valor de la coord en x=', 'coordenada')
    vy = obtener_parametro('dame el valor de la coord en y=', 'coordenada')
    vector = np.array([[vx], [vy]])

    # Selección del modelo de distancia
    print("\nQue modelo deseas usar para calcular las distancias a las clases")
    print("1. Distancia Euclidiana.")
    print("2. Distancia de Mahalanobis.")
    modelo = obtener_parametro("Tu opcion: ", 'modelo')
    
    # Lista de centroides y distancias
    medias = []
    for clase in matrices:
        # mean(..., axis=1) calcula la media por fila y keepdims=True mantiene la forma 2x1
        medias.append(np.mean(clase, axis=1, keepdims=True))

    distancias = np.zeros(num_cl)

    if modelo == 1:
        print("\nHas escogido la distancia Euclidiana.")
        # Calcular la distancia euclidiana: np.linalg.norm(vector)
        for j in range(num_cl):
            distancias[j] = np.linalg.norm(vector - medias[j])
            
    else: # modelo == 2
        print("\nHas escogido la distancia de Mahalanobis.")
        
        # Calcular la distancia de Mahalanobis
        for i in range(num_cl):
            matriz_clase = matrices[i]
            media_clase = medias[i]
            
            # Cálculo de la Matriz de Covarianza
            # np.cov(matriz_clase) calcula la covarianza asumiendo que las observaciones son columnas (axis=1)
            # Para que funcione como MATLAB, debemos transponer la matriz primero: np.cov(matriz_clase)
            matriz_cov = np.cov(matriz_clase)
            
            # Cálculo de la Inversa de la Matriz de Covarianza
            try:
                sigma_inv = np.linalg.inv(matriz_cov)
            except np.linalg.LinAlgError:
                # Si la matriz es singular (no invertible), se usa la pseudo-inversa (Moore-Penrose)
                print(f"Advertencia: Matriz de covarianza de Clase {i+1} es singular. Usando pseudo-inversa.")
                sigma_inv = np.linalg.pinv(matriz_cov) 
            
            # Vector diferencia: (X - mu)
            diff_vector = vector - media_clase
            
            # Distancia de Mahalanobis: sqrt((X - mu)^T * Sigma^-1 * (X - mu))
            # En NumPy/Python: (diff_vector.T @ sigma_inv) @ diff_vector
            distancia_cuadrada = diff_vector.T @ sigma_inv @ diff_vector
            
            # La distancia cuadrada es un array [[valor]], se extrae el valor y se calcula la raíz
            distancias[i] = np.sqrt(distancia_cuadrada[0, 0])

    # --- 4. DETERMINAR LA CLASE MÁS CERCANA ---

    minimo = np.min(distancias)
    clase_asignada = np.argmin(distancias) + 1 
    
    # Umbral de decisión (se usa 400 como en tu código)
    UMBRAL = 250
    
    if minimo > UMBRAL:
        print(f"\nEl vector no pertenece a ninguna clase. Distancia mínima ({minimo:.2f})")
        print(f"es mayor al umbral de {UMBRAL} puntos.")
    else:
        print(f"\nEl vector pertenece a la CLASE: {clase_asignada}")
        print(f"Con una distancia de {minimo:.4f}")

    # --- 5. GRAFICACIÓN ---

    plt.figure(figsize=(10, 8))
    plt.title(f'Clasificación de vectores ({("Euclidiana" if modelo == 1 else "Mahalanobis")})')
    plt.xlabel('Eje X')
    plt.ylabel('Eje Y')
    plt.grid(True)
    
    leyendas = []
    
    # Graficar las clases y los centroides
    for i in range(num_cl):
        x_coords = matrices[i][0, :]
        y_coords = matrices[i][1, :]
        
        # Genera un color RGB aleatorio para cada clase
        color_rgb = np.random.rand(3)

        # Graficar los puntos de la clase
        plt.plot(x_coords, y_coords, 'o', 
                 color=color_rgb, 
                 markersize=8, 
                 markeredgecolor='black',
                 label=f'Clase {i + 1}')
        
        # Graficar el centroide
        plt.plot(medias[i][0, 0], medias[i][1, 0], 'X', 
                 color='k', 
                 markersize=10, 
                 markeredgewidth=2)
                 
        leyendas.append(f'Clase {i + 1}')

    # Graficar el vector desconocido
    plt.plot(vector[0, 0], vector[1, 0], 'P', 
             markersize=12, 
             markerfacecolor='green', 
             markeredgecolor='black',
             label='Vector Desconocido')
             
    leyendas.append('Vector Desconocido')

    # Mostrar la leyenda y la gráfica
    plt.legend(loc='best')
    plt.show()

    # Preguntar al usuario si desea continuar
    try:
        op = int(input('\nDeseas continuar con el programa (-1 para salir): '))
    except ValueError:
        op = -1
        
print('Fin de programa, ahi nos vemos. 👋')